packages/@ember/-internals/routing/lib/services/router.ts

```diff
@@ -393,6 +394,32 @@ export default class RouterService extends Service {
     return this._router._routerMicrolib.recognizeAndLoad(internalURL);
   }
 
+  /**
+   * Refreshes all currently active routes, doing a full transition.
+   * If a routeName is provided and refers to a currently active route,
+   * it will refresh only that route and its descendents.
+   * Returns a promise that will be resolved once the refresh is complete.
+   * All resetController, beforeModel, model, afterModel, redirect, and setupController
+   * hooks will be called again. You will get new data from the model hook.
+   *
+   * @method refresh
+   * @param {String} [routeName] the route to refresh (along with all child routes)
+   * @return Transition
+   * @public
+   */
+  refresh(routeName?: string) {
+    if (!routeName) {
+      return this._router._routerMicrolib.refresh();
+    }
+
+    assert(`The route "${routeName}" was not found`, this._router.hasRoute(routeName));
+    assert(`The route "${routeName}" is currently not active`, this.isActive(routeName));
+
+    let pivotRoute = this._router._routerMicrolib.getRoute(routeName) as Route;
```

Comment:
I'm not sure if this is the best way to get a reference to the route instance.

Reply:
I also added `as Route` since TS used the following type:
```ts
abstract getRoute(name: string): T | Promise<T>;
```

And that isn't compatible with the `refresh` method of `_routerMicrolib` since it expects a `Route`.

Reply:
could you do `getRoute<Route>(routeName)`?

Reply:
@knownasilya That triggers a type error :smile:. `Expected 0 type arguments, but got 1.`

The issue is that the [router.js getRoute method](https://github.com/tildeio/router.js/blob/4f5d411e9ba44efea003bc3159484b4060ebfc2a/lib/router/router.ts#L58) has the following type definition:
`abstract getRoute(name: string): T | Promise<T>;`

While the refresh method [expects a Route type](https://github.com/tildeio/router.js/blob/4f5d411e9ba44efea003bc3159484b4060ebfc2a/lib/router/router.ts#L882).

So without the `as Route` the following type error is shown:
`Argument of type 'Route | Promise<Route>' is not assignable to parameter of type 'Route | undefined'.`

When I look at [the actual getRoute implementation](https://github.com/emberjs/ember.js/blob/22bfcfdac0aeefcf333fb2d6697772934201b43b/packages/%40ember/-internals/routing/lib/system/router.ts#L190) the return type doesn't include the Promise<T> though. So I'm not sure what is up with that, but I have limited TS experience :smile:.